controllers/topolvm_csi_driver.go

```diff
@@ -0,0 +1,99 @@
+
+func (c csiDriver) getName() string {
+	return driverName
+}
+
+func (c csiDriver) ensureCreated(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+	csiDriverDeployment := getCSIDriverDeployment()
+	result, err := cutil.CreateOrUpdate(ctx, r.Client, csiDriverDeployment, func() error {
+		return nil
+	})
+	// CSIDriver resource is an immutable resource and can have result either Created or Unchanged and it's clusterscoped
+	switch result {
+	case cutil.OperationResultCreated:
+		r.Log.Info("csi driver", "operation", result, "name", csiDriverDeployment.Name)
+	case cutil.OperationResultNone:
+		r.Log.Info("csi driver", "operation", result, "name", csiDriverDeployment.Name)
+	default:
+		r.Log.Error(err, "csi driver reconcile failure", "name", csiDriverDeployment.Name)
+		return err
+	}
+	return nil
+}
+
+func (c csiDriver) ensureDeleted(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+	csiDriverDeployment := &storagev1.CSIDriver{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: TopolvmCSIDriverName}, csiDriverDeployment)
+
+	if err != nil {
+		// already deleted in previous reconcile
+		if errors.IsNotFound(err) {
+			r.Log.Info("csi driver deleted", "TopolvmCSIDriverName", csiDriverDeployment.Name)
+			return nil
+		}
+		r.Log.Error(err, "unable to retrieve topolvm csi driver resource", "TopolvmCSIDriverName", csiDriverDeployment.Name)
+		return err
+	}
+
+	// if not deleted, initiate deletion
+	if csiDriverDeployment.GetDeletionTimestamp().IsZero() {
+		if err = r.Client.Delete(ctx, csiDriverDeployment); err != nil {
+			r.Log.Error(err, "unable to delete topolvm csi driver", "TopolvmCSIDriverName", csiDriverDeployment.Name)
+			return err
+		}
+	} else if !csiDriverDeployment.GetDeletionTimestamp().IsZero() {
+		// set deletion in-progress for next reconcile to confirm deletion
+		return fmt.Errorf("topolvm csi driver %s is being uninstalled", csiDriverDeployment.Name)
+	}
+
+	return nil
+}
+
+func (c csiDriver) updateStatus(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+	// intentionally empty as there'll be no status field on CSIDriver resource
+	return nil
+}
+
+func getCSIDriverDeployment() *storagev1.CSIDriver {
+	// topolvm doesn't use/need attacher and reduce a round trip of the rpc by setting this to false
+	attachRequired := false
+	podInfoOnMount := true
+
+	// use storageCapacity tracking to take scheduling decisions
+	storageCapacity := true
+	csiDriver := &storagev1.CSIDriver{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: TopolvmCSIDriverName,
+		},
+		Spec: storagev1.CSIDriverSpec{
+			AttachRequired:  &attachRequired,
+			PodInfoOnMount:  &podInfoOnMount,
+			StorageCapacity: &storageCapacity,
+			// topolvm supports both Ephemeral and Persistent volumes
+			VolumeLifecycleModes: []storagev1.VolumeLifecycleMode{storagev1.VolumeLifecyclePersistent, storagev1.VolumeLifecycleEphemeral},
```

Comment:
Do we need storagev1.VolumeLifecycleEphemeral here?

Reply:
- it's really beneficial in throw off containers and for some workloads which just needs storage to save logs which are of no use and need not to be persistent (let's say initialisation logs of an nginx)
- topolvm supports this with ease and tested, IMHO it's better to have this

Reply:
Since we are not supporting this for 4.10, I think we should remove it. We can add it in a future release.